tests/std/tests/P0798R8_monadic_operations_for_std_optional/test.cpp

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+using namespace std;
+
+struct Immovable {
+    constexpr Immovable(int x) : value(x) {}
+    Immovable(const Immovable&) = delete;
+    Immovable(Immovable&&)      = delete;
+    Immovable& operator=(const Immovable&) = delete;
+    Immovable& operator=(Immovable&&) = delete;
+    constexpr ~Immovable() {}
+
+    int value;
+};
+
+template <class T>
+struct fn {
+    template <class U>
+    constexpr auto operator()(U&&) & {
+        static_assert(is_same_v<T, U&&>);
+        return optional<int>{48};
+    }
+
+    constexpr auto operator()() && {
+        return optional<long>{49};
+    }
+
+    template <class U>
+    constexpr auto operator()(U&& u) && {
+        static_assert(is_same_v<T, U&&>);
+        return Immovable{u + 2};
+    }
+};
+
+template <class Optional>
+constexpr void test_impl(Optional&& nonempty, Optional&& empty) {

```

Comment:
Naming nitpick: `empty` is the name of a Standard function and I generally recommend against avoiding potential shadowing. Occurs below.